[FP][17.0][IMP] web_timeline - support multiple group-by levels and allow to group by m2m field

cuongnmtm opened the pull request:
Forward-port of #3006

Test cases:

- [x]  group by one field (group by m2o, selection, integer, m2m, date, ...)
- [x]  group by multiple fields
- [x]  group by 0 field: back to default group by
- [x]  create a new record
- [x]  drag and drop record to update
- [x]  click on the group by to view the group by record detail
- [x]  when group by m2m field, record is displayed multiple times if linked to multiple m2m records
- [x]  when drag and drop, the value of the m2m field should not be updated as there is no best way to handle this
- [x]  when creating a new record, set the default value for all grouped fields
  - [x] m2m field value will also be set, starting from 17.0
- [x]  handle group by non-relation fields: click on the group of selection or char field do not popup record form view to avoid error.
- [x]  dependency_arrow between multiple collapse/expand groups
- [x] group by create_date year|month|day, try colapse the group by (year) in both project and task timelines

carlos-lopez-tecnativa commented:
@cuongnmtm I understand this PR is not ready for review yet because the files are not loaded in the manifest.
Additionally, you must convert them to .esm.js and odoo-module according to this commit made during the migration to 17.0: https://github.com/OCA/web/commit/43341291dd0a61875fb4e45e75baff096420ac5e

cuongnmtm committed:
[IMP] web_timeline - support multi group by level

carlos-lopez-tecnativa reviewed (changes requested):
I got this error when grouping a datetime field. I suspect this commit https://github.com/OCA/web/pull/3006/commits/751fb9c3699f6926cc453871c8d83035f3ef6d8b might be missing, please review: 

<img width="1372" height="693" alt="image" src="https://github.com/user-attachments/assets/6d7c80f7-a78f-4694-aa3f-d024996c6328" />

cuongnmtm commented:
> I got this error when grouping a datetime field. I suspect this commit [751fb9c](https://github.com/OCA/web/commit/751fb9c3699f6926cc453871c8d83035f3ef6d8b) might be missing, please review:
> 
> <img alt="image" width="1372" height="693" src="https://private-user-images.githubusercontent.com/174346082/498855052-6d7c80f7-a78f-4694-aa3f-d024996c6328.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjAwMzIyMDYsIm5iZiI6MTc2MDAzMTkwNiwicGF0aCI6Ii8xNzQzNDYwODIvNDk4ODU1MDUyLTZkN2M4MGY3LWE3OGYtNDY5NC1hYTNmLWQwMjQ5OTZjNjMyOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAwOVQxNzQ1MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lODlhZTU1NjA5N2I0ZjJiNjFlY2JmYzVhMDZiMWRhNTdlNDY3MGM1ZTZjNWZiOGEwYTcxYjc3MDQxYWY0MGFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.MC37lIP3HNOH2Rxx4cOVcEEb1ornygwUckG9BvfSznw">

I haven't finished this port yet. There are three commits to port, and the remaining two are taking more time because I need to adapt the changes to the refactoring in version 17. I will find time to work on them.

pedrobaeza commented:
@cuongnmtm any news on this fw-port?

cuongnmtm commented:
> @cuongnmtm any news on this fw-port?

@pedrobaeza Thank you for checking this. I have reviewed it and believe it will take several hours to complete, as porting this change involves significant effort due to the major refactoring in 17.0.

I am currently quite busy with the Komit's customer projects, but I will aim to continue working on this every Friday.

pedrobaeza commented:
Thanks for the follow-up. We will wait for news.

cuongnmtm committed:
[IMP] web_timeline - support group by date/datetime fields by year|quarter|month|week|day

Port of 16.0 commit 751fb9c to 17.0 refactored codebase:
- Convert group IDs from string format to JSON segment objects
- Add _getGroupedDate() for date/datetime grouping with Luxon
- Transform date fields after search_read for :operator specifiers
- Handle selection fields to display labels instead of technical values
- Remove fieldsGet dependency, use model.fields directly
- Add M2M name caching in split_groups
- Update controller group parsing (click, move, add) for JSON format

cuongnmtm committed:
[IMP] web_timeline - set m2m group value as default when creating record

When creating a new record from a m2m group row, pre-fill the m2m field
with the group's value using the [(6, 0, [id])] command format.

cuongnmtm commented:
It is done, could you please help review again @pedrobaeza @carlos-lopez-tecnativa ? Thanks.

pedrobaeza reviewed (approved):
Tested on runboat and working properly. Thanks!

carlos-lopez-tecnativa commented on web_timeline/static/src/views/timeline/timeline_renderer.esm.js:
```diff
@@ -343,58 +373,277 @@ export class TimelineRenderer extends Component {
+            if (m2mNameCache[cacheKey]) {
+                return m2mNameCache[cacheKey];
+            }
+            const names = await this.orm.call(model, "name_get", [ids]);
```
I think you can use `display_name` instead of `name_get`, considering the forward-port to V18.

cuongnmtm replied:
Thanks for the display_name suggestion, I will update the code.

For the issue related to the create_date group by, I don't face the issue in localhost; perhaps a regeneration of assets is required?

<img width="1456" height="706" alt="image" src="https://github.com/user-attachments/assets/e2a3cbfa-64d8-434e-b201-14aff1eb07bc" />

carlos-lopez-tecnativa replied:
> Thanks for the display_name suggestion, I will update the code.
> 
> For the issue related to the create_date group by, I don't face the issue in localhost; perhaps a regeneration of assets is required?
> 

I noticed you tested it in **projects**, but I tested it in **tasks**. I attached a video explaining the error; this was tested on a fresh database. I imagine this occurs in this model because it has **dependencies**, whereas in the project it does not.


https://github.com/user-attachments/assets/db635a60-6b3e-47f7-a62f-99b0163e04af

cuongnmtm replied:
Nice catch and thanks for the video!

I have fixed it.

carlos-lopez-tecnativa reviewed (changes requested):
I have the same issue reported in V16 https://github.com/OCA/web/pull/3006#pullrequestreview-3168525063

I tested locally with the `project_timeline` module installed. In the `All Tasks` view, when grouping by `creation date` (by year and month), collapsing the group by year triggers a traceback.
<img width="1321" height="811" alt="Image" src="https://github.com/user-attachments/assets/e7664f2c-cb66-4660-bf5e-c5bd89153e83" />

carlos-lopez-tecnativa reviewed (approved):
Nice work, LGTM. Just a minor comment: I think you should update the README, because now the group is opened with a double-click, but the README says it opens with a single click.
<img width="1286" height="663" alt="Image" src="https://github.com/user-attachments/assets/f0a8cfc3-f8ca-41df-9fa3-ccf673971429" />

cuongnmtm commented:
> Nice work, LGTM. Just a minor comment: I think you should update the README, because now the group is opened with a double-click, but the README says it opens with a single click. <img alt="Image" width="1286" height="663" src="https://private-user-images.githubusercontent.com/174346082/569753064-f0a8cfc3-f8ca-41df-9fa3-ccf673971429.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ2MDA3MzEsIm5iZiI6MTc3NDYwMDQzMSwicGF0aCI6Ii8xNzQzNDYwODIvNTY5NzUzMDY0LWYwYThjZmMzLWY4Y2EtNDFkZi05ZmEzLWNjZjY3Mzk3MTQyOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyN1QwODMzNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYmZhZDRlMTg0MTUyMDI0MTU5ZjYzMDg0MmEyOGMyMmYyYjNiMjIzOGExMTAxNjc1NTZiOWM1MjQ3ZjdiZjY1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.lb6jU6bUUGvTyFMzviPNLAi-b1xjCOjfWvlipuHYN74">

I updated the README. Thank you.

pedrobaeza commented:
Thank you very much for the efforts!

/ocabot merge minor

Are you going to fw-port it to 18?

cuongnmtm commented:
> Thank you very much for the efforts!
> 
> /ocabot merge minor
> 
> Are you going to fw-port it to 18?

Yes sir, I am going to initialize the fw. Hopefully, it is easier forwarding to 18.0 and 19.0.

paradoxxxzero commented:
> > Thank you very much for the efforts!
> > /ocabot merge minor
> > Are you going to fw-port it to 18?
> 
> Yes sir, I am going to initialize the fw. Hopefully, it is easier forwarding to 18.0 and 19.0.

FYI: #3518
